Added auto off timer feature - Garage door support

handy-manny-swe opened the pull request:
Fixes https://github.com/mongoose-os-apps/shelly-homekit/issues/11

Tested it during this weekend on my garage door with success.

This is my first go at C for 20 years, so a critical eye might be a good idea. 

Cheers!
Johan

andyblac commented:
why a max timeout of 5 mins ?, why have a max timeout at all, this would come in handy for auto turn off lights that kids/partner leaves on, ie could be set to 20mins so if left on light would auto turn off after 20mins.

could you remove the max timeout and possibly add an option for secs or minuets ?

also you forgot to bump the Version number.

handy-manny-swe commented:
> why a max timeout of 300 seconds (5 mins) ?, why have a max timeout at all, this would come in handy for auto turn off lights that kids/partner leaves on, ie could be set to 20mins so if a light is left on it would auto turn off after 20mins.
> 
> could you remove the max timeout and possibly add an option for secs or minuets ?
> 
> also you forgot to bump the Version number.

Appreciate the feedback, and good points! 

I have removed the upper limit and changed the input to allow hours, minutes and seconds, and added some basic validation for the input.
Also bumped the version number.

andyblac commented:
thanks for taking on board my comments, really appreciate it.

rojer commented on src/shelly_sw_service.c:
```diff
@@ -100,6 +116,25 @@ static void shelly_sw_set_state_ctx(struct shelly_sw_service_ctx *ctx,
     }
   }
+
+  if (cfg->auto_off && auto_off_timer_id >= 0)
```
wrap this whole section in `if (cfg->auto_off)`, this will make other conditions simpler.

rojer commented on src/shelly_sw_service.c:
```diff
@@ -100,6 +116,25 @@ static void shelly_sw_set_state_ctx(struct shelly_sw_service_ctx *ctx,
+
+static void do_auto_off(void *arg) {
+  struct shelly_sw_service_ctx *ctx = arg;
+  shelly_sw_set_state_ctx(ctx, false, "auto_off");
```
check for cfg->auto_off again here, it's possible that mode is changed while the timer is running.

also set `s_auto_off_timer_id` to `MGOS_INVALID_TIMER_ID`, since timer id for one-shot timer is invalidated when it fires.

rojer commented on src/shelly_sw_service.c:
```diff
@@ -51,6 +51,10 @@ struct shelly_sw_service_ctx {
 
 static struct shelly_sw_service_ctx s_ctx[NUM_SWITCHES];
 
+static int auto_off_timer_id = -1;
```
MGOS_INVLAID_TIMER_ID instead of -1, rename to s_auto_off_timer_id

rojer commented on src/shelly_sw_service.c:
```diff
@@ -67,6 +71,17 @@ static void do_reset(void *arg) {
 
+static void reset_auto_off_timer() {
+  mgos_clear_timer(auto_off_timer_id);
+  auto_off_timer_id = -1;
```
ditto

rojer commented on mos.yml:
```diff
@@ -29,6 +29,8 @@ config_schema:
   - ["sw.state", "b", false, {"State of the switch"}]
   - ["sw.persist_state", "b", false, {"Whether state of the switch should be persisted across reboots."}]
+  - ["sw.auto_off", "b", false, {"Whether the switch should automatically turn OFF after turning ON"}]
+  - ["sw.auto_off_delay", "s", "", {"Delay for automatically turning OFF, in the 24 hour format HH:mm:ss."}]
```
i'd rather we keep string conversion operations in JS and make this an int variable

andyblac commented:
forget to version bump ;)

handy-manny-swe commented:
> first of all, thanks for doing this!
> see my review comments.
> also please rebase onto latest master.

Thanks for providing this excellent repo to start from! I'm glad that I can give something back.

I think I've covered your comments now.

rojer commented on src/shelly_sw_service.c:
```diff
@@ -100,6 +124,16 @@ static void shelly_sw_set_state_ctx(struct shelly_sw_service_ctx *ctx,
+  s_auto_off_timer_id = MGOS_INVALID_TIMER_ID;
+  struct shelly_sw_service_ctx *ctx = arg;
+  const struct mgos_config_sw *cfg = ctx->cfg;
+  if (cfg->auto_off)  // Don't set state if auto off has been disabled during timer run
```
please add braces here, it's a requirement for multiline conds for maximum clarity.

handy-manny-swe replied:
it's just a comment, but sure, I've added brackets

handy-manny-swe commented:
> almost there. please run clang-format on the modified files to apply style (set in .clang-format)

Done!

andyblac commented:
any idea when this will be pulled ?

rojer commented on src/shelly_main.c:
```diff
@@ -304,14 +304,16 @@ static void shelly_get_info_handler(struct mg_rpc_request_info *ri,
 #ifdef MGOS_CONFIG_HAVE_SW1
-      "sw1: {id: %d, name: %Q, in_mode: %d, persist: %B, state: %B"
+      "sw1: {id: %d, name: %Q, in_mode: %d, persist: %B, state: %B, auto_off: "
+      "%B, auto_off_delay: %d}"
```
there shouldn't be closing brace at the end here

rojer replied:
"sw1: {id: %d, name: %Q, in_mode: %d, persist: %B, state: %B, "
      "auto_off: %B, auto_off_delay: %d"

handy-manny-swe replied:
Oops! Good find, fixing...

rojer commented on src/shelly_main.c:
```diff
@@ -304,14 +304,16 @@ static void shelly_get_info_handler(struct mg_rpc_request_info *ri,
 #ifdef MGOS_CONFIG_HAVE_SW2
-      "sw2: {id: %d, name: %Q, in_mode: %d, persist: %B, state: %B"
+      "sw2: {id: %d, name: %Q, in_mode: %d, persist: %B, state: %B, auto_off: "
+      "%B, auto_off_delay: %d}"
```
ditto

rojer replied:
"sw2: {id: %d, name: %Q, in_mode: %d, persist: %B, state: %B, "
      "auto_off: %B, auto_off_delay: %d"

handy-manny-swe replied:
Ditto :)

rojer commented:
merged, thank you!

andyblac commented:
just letting you know it was not bumped to version 1.8.0

rojer commented:
yeah, i know. i'll hold on releasing a bit, i'm thinking of doing #22 as well and pushing together.

andyblac commented:
ok, np, just let you know I built myself and test seems to work pretty well.

rojer commented:
released in 1.8.0

Flying150 commented:
Hello Deomid,

This is awesome! Thanks for this release ;o)
I have a specific application that would require pulses in Milli second steps.
The regular firmware can be set up this way. Any chance that the next release could consider this format: e.g.: 00:00:00:01 or 00:00:00:05 ?

Thanks for your support in advance,

Martin Laszák

> Am 19.04.2020 um 00:44 schrieb Deomid Ryabkov <notifications@github.com>:
> 
> ﻿
> released in 1.8.0
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
